README.md
# geouy <img align="right" src="man/figures/geouy_logo_a.png" alt="logo" width="160"> <img align="right" src="man/figures/geouy_logo_b.png" alt="logo" width="160">

<!-- badges: start -->

[![CRAN/METACRAN Version](https://www.r-pkg.org/badges/version/geouy)](https://CRAN.R-project.org/package=geouy)
[![CRAN/METACRAN Total downloads](https://cranlogs.r-pkg.org/badges/grand-total/geouy?color=blue)](https://CRAN.R-project.org/package=geouy) 
[![CRAN/METACRAN downloads per month](https://cranlogs.r-pkg.org/badges/geouy?color=orange)](https://CRAN.R-project.org/package=geouy)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![R build status](https://github.com/RichDeto/geouy/workflows/R-CMD-check/badge.svg)](https://github.com/RichDeto/geouy/actions)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/RichDeto/geouy?branch=master&svg=true)](https://ci.appveyor.com/project/RichDeto/geouy)
[![Codecov test coverage](https://codecov.io/gh/RichDeto/geouy/branch/master/graph/badge.svg)]( https://app.codecov.io/gh/RichDeto/geouy?branch=master)
[![DOI](https://zenodo.org/badge/229800365.svg)](https://zenodo.org/badge/latestdoi/229800365)

<!-- badges: end -->

**geouy** is an R package that allows users to easily access official spatial data sets of Uruguay. The package includes a wide range of geospatial datasets as *simple features* (`sf`), available at various geographic scales and for various years with harmonized attributes and projection (see detailed list below).

## Installation

```R
# From CRAN
  install.packages("geouy")
  library(geouy)
  
# Use the development version with latest features
  utils::remove.packages('geouy')
  devtools::install_github("RichDeto/geouy")
  library(geouy)
```
obs. If you use **Linux**, you need to install a couple dependencies before installing the libraries `sf` and `geouy`. [More info here](https://github.com/r-spatial/sf#linux). 


## Basic Usage

### load_geouy

The syntax of all `geouy` functions operate one the same logic so it becomes intuitive to download any data set using a single line of code. Like this:

```R
secc <- load_geouy("Secciones")
```

#### Available datasets:


##### Administrative limits 

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"Uruguay"`| `"INE"` | MIDES | 2011 | wfs |
|`"Areas administrativas"`| `"SGM"` | SGM | 2011 | wfs |
|`"Deptos"`| `"INE"` | IDE | 2011 | wfs |
|`"Dptos"`| `"DINAMA"` | "MVOTMA"| 2020  | wfs | 
|`"Limites departamentales"`| `"IGM"` | "IGM" | 2011  | wfs | 
|`"Departamentos"`| `"IDE"` | "MIDES" | 2011  | wfs |
|`"Secciones"`| `"INE Censo"` | MIDES | 2011 | wfs |
|`"Segmentos"`| `"INE Censo"` | MIDES | 2011 | wfs |
|`"Zonas"`| `"INE Censo"` | MIDES | 2011 | wfs |
|`"Secc MVD 2004"`| `"INE"` | MIDES | 2004 | wfs |
|`"Segm MVD 2004"`| `"INE"` | MIDES | 2004 | wfs |
|`"Segm URB INT 2004"`| `"INE"` | MIDES | 2004 | wfs |
|`"Zonas MVD 2004"`| `"INE"` | MIDES | 2004 | wfs |
|`"Zonas URB INT 2004"`| `"INE"` | MIDES | 2004 | wfs |
|`"Localidades pg"`| `"INE Censo"` | MIDES | 2011 | wfs |
|`"Localidades pt"`| `"INE Censo"` | MIDES | 2011 | wfs |
|`"Centros poblados pg"`| `"SGM"` | SGM | 2011 | wfs |
|`"Centros poblados pt"`| `"SGM"` | SGM | 2011 | wfs |
|`"Municipios10"`| `"DINOT-IM-IC"` | MVOTMA | 2010 | zip |
|`"Municipios15"`| `"DINOT-IM-IC"` | MVOTMA | 2015 | zip |
|`"CCZ"`| `"INE"` | MIDES | 2011 | wfs |
|`"Asentamientos irregulares"`| `"PMB"` | MIDES | 2014 | wfs |
|`"Barrios"` | `"INE"` | MIDES | 2011 | wfs |
|`"Balnearios"`| `"MTOP"` | MTOP |  2017 | wfs |
|`"Secciones catastrales"` | `"DNC"` | MVOTMA | 2013 | zip |
|`"Padrones rurales"` | `"DNC"` | "MVOTMA" | 2014 | zip |  
|`"Padrones urbanos"`| `"DNC"` | MVOTMA | 2014 | zip | 
|`"Secciones policiales"` | `"MI"` | MVOTMA | 2017 | zip |

##### Demography

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"LocHog11"`| `"INE"` | IDE | 2011 | zip |
|`"LocPobHom11"`| `"INE"` | IDE | 2011 | zip |
|`"LocPobMuj11"`| `"INE"` | IDE | 2011 | zip |
|`"LocViv11"`| `"INE"` | IDE | 2011 | zip |


##### Hidrology

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"Cuencas hidro N1"`| `"DINAGUA"` | MVOTMA | 2020 | zip |
|`"Cuencas hidro N2"`| `"DINAGUA"` | MVOTMA | 2020 | zip |
|`"Cuencas hidro N3"`| `"DINAGUA"` | MVOTMA | 2020 | zip |
|`"Cuencas hidro N4"`| `"DINAGUA"` | MVOTMA | 2020 | zip |
|`"Cuencas hidro N5"`| `"DINAGUA"` | MVOTMA | 2020 | zip |
|`"Cursos de agua navegables y flotables"` | `"MTOP"` | MTOP | 2019 | wfs |
|`"Lagunas publicas"` | `"MTOP"` | MTOP | 2019 |  wfs |
|`"Ambientes acuaticos"` | `"FREPLATA"` | MVOTMA | 2009 | zip |
|`"Areas protegidas"` |`DINAMA"` | MVOTMA | 2015 | zip |
|`"Baniados"` | `"DINAMA"` | MVOTMA | NA | zip |
|`"Batimetria"` | `"DINAMA"` | MVOTMA | 2020 | zip | 


##### Ways

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"Rutas"` | `"IDE"` | MIDES | 2017 | wfs |
|`"Calles"` | `"IDE - UTE - IM"` | MIDES | 2017 | wfs |
|`"Peajes"` | `"MTOP"` | MTOP | 2019 | wfs |
|`"Postes Kilometros"` |`"MTOP"` | MTOP | 2019 | wfs |

##### Services

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
| `"OTs"` | `"MIDES"` | MIDES | 2022 | zip |
| `"Educación en Primera Infancia e Inicial"` | `"CEIP"` | MIDES | 2020 | zip | 
| `"Jardines de infantes"` | `"CEIP"` | MIDES | 2020 | zip |
| `"Colegios privados N0a3"` | `"CEIP"` | MIDES | 2020 | zip |
| `"Escuelas con N3"` | `"CEIP"` | MIDES | 2020 | zip |
| `"Escuelas"` | `"CEIP"` | MIDES | 2020 | zip | 
| `"Educacion especial"` | `"CEIP"` | MIDES | 2020 | zip | 
| `"Educacion secundaria"` | `"CEIP"` | MIDES | 2020 |  zip |
| `"UTU"` | `"ANEP"` | MIDES | 2020 |  zip |
| `"Instituciones deportivas"` | `"IDE"` | MIDES | 2015 | wfs |
| `"Playas"`| `"DINAMA¨` | MVOTMA | 2007 | zip |

##### Orthophotos

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"Grilla ortofotos urbana"` | `"IDE"` | IDE | 2019 | wfs |
|`"Grilla ortofotos nacional"` | `"IDE"` | IDE | 2019 | wfs |
  
##### Land Cover

| Layer | Productor | Source | Year | Format |
|-----|-----|-----|-----|-----|
|`"Cobertura suelo 2000"` | `"DINAGUA"` | MVOTMA | 2000 | zip |
|`"Cobertura suelo 2008"` | `"DINAGUA"` | MVOTMA | 2008 | zip |
|`"Cobertura suelo 2011"` | `"DINAGUA"` | MVOTMA | 2011 | zip |
|`"Cobertura suelo 2015"` | `"DINAGUA"` | MVOTMA | 2015 | zip |
|`"CONEAT"` | `"RENARE"` | "RENARE" | NA  | wfs |



### which_uy() 

Add to an `sf` object its spatial coincidence with one or more administrative units in Uruguay, generating the corresponding variables.

### where_uy() 

You get an `sf` object of one or more administrative units in Uruguay, according to a query by code or name in the layer.

### add_geom()

This function allows you to add a `geom` variable with a code variable of "zona", "barrio", "localidad", "segmentos", "secciones" or "departamentos".

### geocode_ide_uy() 

Allows geocoding directions using IDE_uy.

### plot_geouy()

Plot a variable of your `sf` object with north and scale, set on a simple theme.

### Other functions:

`is.uy4326():` Test if an `sf` object match with Uruguay at crs = 4326.
`is.uy32721():` Test if an `sf` object match with Uruguay at crs = 32721.
`is.uy5381():` Test if an `sf` object match with Uruguay at crs = 5381.
`is.uy5382():` Test if an `sf` object match with Uruguay at crs = 5382.

## Datasets

### metadata

This dataset has the metadata of all vector geometries provided by `geouy` and detailed at the top

### metadata_wms

This dataset has the metadata of all the raster services that will be provided by `geouy`, although the functions are still under development.

### loc_agr_ine

This dataset allows you to aggregate Localidades INE by the aglomerations used by INE. An example of use can be: 

``` r
loc <- which_uy(base, "Localidades pg") %>% 
    dplyr::left_join(loc_agr_ine, by = c("cod_Localidades pg" = "codloc"))
```

### uy_deptos_grid and mvd_barrios_grid 

Two datasets to use as `geofacet` grid dataset for departments of Uruguay and neighborhoods of Montevideo.


## History

This package arises from the conjugation of own ideas with an eye on the region. It started as a part of the package where I work with @calcita at [ech](https://github.com/calcita/ech), and some geospatial service packages in the region mainly: 
[geobr](https://github.com/ipeaGIT/geobr) and 
[chilemapas](https://github.com/pachadotdev/chilemapas)

This walk on the shoulders of giants, allows this package focused on this small country (my beautiful Uruguay), to have its own particularities although it tries to fit especially to [geobr](https://github.com/ipeaGIT/geobr) in its structure and with a view to complementing [ech](https://github.com/calcita/ech).

## Community contributions [es](https://github.com/RichDeto/geouy/issues/1)

This package intends to incorporate any function of general requirements that use the geographic data of Uruguay as a base. All contributions in this regard are welcome.

If you work with geographic data of Uruguay and want to add your function or data, we recommend that you read the following tips on how to collaborate:

### To add your function:

    - Fork of this repository
    - Add your function as an `.R` file in the `R/` folder with the same name as the function
    - Document it with `roxygen2` clarifying its functionality, parameters and an example of use. To see how it is documented to review another function in the same directory
    - Remember at the end of the documentation you must add a # '@export
    - Check that the types and values of your function parameters are fine (for example, you can look at `data-raw/metadata.R`
    - Add the dependencies in the `DESCRIPTION` file 
    - Check the package with devtools::check()
    - If everything works fine, then make a pull request

### To add geographic data to the `load_geouy()` function:

    - Fork of this repository
    - Identify the corresponding WFS service URL
    - Add a record to the `metadata` file in the `data-raw/` directory, with the corresponding data.
    - Also include this record in the corresponding table of the `README.md` file, with the corresponding format.
    - If everything works fine, then make a pull request

## <i class="fa fa-pencil" aria-hidden="true"></i> Citation

To cite `geouy` in publications, please use:

Detomasi, Richard (2021) "geouy: Geographic Information of Uruguay". R package version 0.2.5
   URL: https://github.com/RichDeto/geouy. 
  
A BibTeX entry for LaTeX users is:

```
@Misc{geouy,
  title = {geouy: Geographic Information of Uruguay},
  author = {Richard Detomasi},
  note = {R package version 0.2.5},
  year = {2021},
  url  = {https://github.com/RichDeto/geouy},
}

```


NEWS.md
# log history of geouy package development

-------------------------------------------------------

## geouy v0.2.8 (2023-08-22)

* update geouy.R for changes in roxygen2
* debug warning in add_geom.R

## geouy v0.2.7 (2023-05-16)

* Debug testing errors
* update plot_geouy for deprecated ggplot2::aes_string 
* update which_uy for changes in tidyselect 0.1.1

## geouy v0.2.6 (2022-10-04)

* Add Municipios 2010 and 2015, and Territorial Offices of MIDES to metadata 
* Update geocode_ide_uy() because server of IDEuy change
* Add a reverse_ide_uy() function for reverse geocoding 
* Change geoservice of "Departamentos"
* Change code variable name of "Municipios"
* Update URLs 

## geouy v0.2.5 (2021-08-12)

* Add demographic links from IDEuy
* Add an encoding variable to metadata.
* Implementation of encoding definition to load_geouy()
* Add a uy_deptos_grid and a mvd_barrios_grid as geofacet grid dataset 
* Improve citacion year and version

## geouy v0.2.4 (2021-05-02)

* plot_goeuy % bug fixed
* tutorial
* Add wfs for CONEAT of RENARE
* Update links of INE wfs, connecting to "Geoportal del Instituto Nacional de Estadística" made by IDE-Uy
* Update links of DINAMA wfs, because institutional changes make a change in server name
* Add metadata_tables for tables with geocodings
* Add to metadata cod variables for Padrones
* Add verification of internet access

## geouy v0.2.3 (2020-09-15)

* Add C++11 system requirement for sf update.
* Add CCZ datasets to metadata.
* Add a where_uy() to get an sf object by a name or id consult. 
* Add a add_geom() to add a geom variable to a data.frame by a link variable.
* Improve plot_geouy() fixing bugs, and made compatible with ech:: library.


## geouy v0.2.2 (2020-07-31)

* Add loc_agr_ine dataset of "Localidades agragadas" of INE.
* Add a geouy.R to discribe the package and set globals variables
* replace ggsn with ggspatial to improve north and scale aesthetics in plot_geouy 
* remove lat and long axis in plot_goeuy
* Add a param labels in plot_geouy for labels posiblity, if "%" porcentage with 1 decimal labels, if "n" the value is the label, if "c" put other variable in other_lab

## geouy v0.2.1 (2020-06-09)
 
* Add educational layers to load_geouy(): "Colegios privados N0a3", "Educación en Primera Infancia e Inicial",  "Jardines de infantes", "Escuelas",   "Escuelas con N3", "Educacion especial", "Educacion secundaria" y "UTU".
* Add tryCatch for download.file() in load_geouy() to diferents zip files.

## geouy v0.2.0 (2020-05-07)

* Verify compatibility with 1.0
* Change the use of to RCurl::getURL at geocode_ide_uy()
* Try change tiles_ide_uy() from raster to terra, but at the crop step the ptr slot is lost.
* Improvement of plot_geouy
* Remove tiles_ide_uy() function, because IDEuy change require methods


## geouy v0.1.9 (2020-03-20)

* Update test with CRS structure for sf 0.9 version
* Limits in geocode_geouy(), you must be part of uruguayan public organism and  fill this (forms)[https://www.gub.uy/agencia-gobierno-electronico-sociedad-informacion-conocimiento/comunicacion/publicaciones/formularios-alta-consumo-servicios-pdi] if your organism is not yet vinculated.


## geouy v0.1.8 (2020-03-17)

* Add to 'tiles_ide_uy()' function options to download .jpg in addition to .tif
* Add more testthat for 'tiles_ide_uy()' function
* Add zip format to 'load_goeuy()'
* Add MVOTMA datasets to metadata, i.e.: Ambientes acuaticos, Areas protegidas, Batimetria, Secciones catastrales, Padrones rurales y urbanos, Secciones policiales, Playas and Cuencas hidrograficas in its 5 nivels.
* Add to tiles_ide_uy the posibility for Montevideo tiles with urban = TRUE give orthophotos with 10cm per pixel 
* Solved bug with raster::mosaic for combination of multiple tiles to cover a bbox.
* Add a complementary function for 5382 crs evaluation

## geouy v0.1.7 (2020-02-26)

* Add layers for IDEuy orthophotos grids at 'load_geouy()' 
* Add a 'tiles_ide_uy()' function and it's testthat dile
* Add a complementary function for 5381 crs evaluation
* Change all 'paste()' with 'glue()'
* Add to testthat files 'skip_on...()' functions for a faster evaluation.

## geouy v0.1.6 (2020-02-09)

* Add \value to .Rd files
* Add all variables option to which_uy function
* Complementary functions for the evaluation of typical Uruguayan CRS codes (4326 and 32721)

## geouy v0.1.5 (2020-01-24)

* omit the redundant part "Toolbox for" in your title.
* web reference for the data source/used API in Description 
* write package names, software names and API names in single quotes.
* replace \dontrun{} by \donttest{} in Rd-files 
* Added in metadata the links to "Secc MVD 2004", "Segm MVD 2004", "Segm URB INT 2004", "Zonas MVD 2004", "Zonas URB INT 2004", "Localidades pt", "Instituciones deportivas"

## geouy v0.1.4 (2020-01-22)

 * New function `which_uy()` which allows to add to an "sf" object its spatial coincidence with one or more administrative units in Uruguay, generating the corresponding variables.
 * Added in metadata the links to "Departamentos" and "Barrios" of the INE, renaming "Departamentos" from IGM to "Limites departamentales"
 * testthat for the new function.

## geouy v0.1.3 (2020-01-17)

  * Change readLines for xml2::read_html in geocode_ide_uy()
  * Incorporate Viridis colors to 'plot_geouy()'
  * Added testthat function to all functions   


## geouy v0.1.1 (2020-01-15)

  * After first release of it to CRAN R, details were corrected in the package documentation
  * Was added checks in Travis (linux and mac), Appveyor and Codecov 
  * Generated a DOI in Zenodo 


## geouy v0.1.0 (2019-12-30)

Launch of **geouy** v0.1.0 on [GitHUb](https://github.com/RichDeto/geouy) with the following data sets:    
  * Localidades_pg    
  * Secciones    
  * Asentamientos irregulares    
  * Balnearios    
  * Departamentos    
  * Areas administrativas
  * Segmentos
  * Zonas
  * Centros poblados pg
  * Centros poblados pt
  * Cursos de agua
  * Lagunas
  * Rutas
  * Calles

R/add_geom.R
#' This function allows you to add a geom variable with a code variable of "zona", "barrio", "localidad", "segmentos", "secciones" or "departamentos".
#' @family service
#' @param data data.frame
#' @param unit spatial unit of data, may be: "Departamentos", "Secciones", "Secc MVD 2004", "Segmentos", "Segm MVD 2004", "Segm URB INT 2004", "Zonas", "Zonas MVD 2004", "Zonas URB INT 2004", "Localidades pg", "Municipios" o "Barrios".
#' @param variable Variable name of unit code (without duplicates)
#' @param crs Coordinates Refence Sistem, usually in region 32721 or 4326 (default 32721)
#' @importFrom glue glue
#' @importFrom rlang .data
#' @importFrom dplyr left_join select rename mutate
#' @return data.frame
#' @details
#' Disclaimer: This script is not an official INE product.
#' Aviso: El script no es un producto oficial de INE.
#' @export
#' @examples
#' \donttest{
#' pobre_x_dpto <- as.data.frame(cbind(nomdpto = c("ARTIGAS", "DURAZNO", "FLORIDA", "LAVALLEJA"),
#'                Pobreza = c(0.26, 0.27, 0.07, 0.10)))
#' pobre_x_dpto_geo <- add_geom(data = pobre_x_dpto, unit = "Deptos", variable = "nomdpto")
#' }

add_geom <- function(data, unit, variable, crs = 32721){
  
  # checks ---
  assertthat::assert_that(is.data.frame(data))
  assertthat::assert_that(unit  %in% c("Departamentos", "Deptos", "Secciones", "Secc MVD 2004", "Segmentos", "Segm MVD 2004", "Segm URB INT 2004", "Zonas",
                                       "Zonas MVD 2004", "Zonas URB INT 2004", "Localidades pg", "Municipios", "Barrios"),
                          msg =  glue::glue("Sorry... :( \n {unit} is not a valid value"))
  assertthat::assert_that(variable  %in% names(data), msg =  glue::glue("Sorry... :( \n {variable} is not in data"))
  assertthat::assert_that(sum(duplicated(data[,variable])) == 0, msg =  glue::glue("Sorry... :( \n {variable} in data have duplicated values"))
  
  g <- geouy::load_geouy(unit, crs)
  md <- geouy::metadata
  
  if (is.character(data[[variable]])) {
    name <- as.character(md[md$capa == unit, "name"])
    g2 <- g %>% dplyr::select(all_of(name)) %>%
      dplyr::rename("link" = name) %>% 
      dplyr::mutate(link = as.character(link))
    data <- dplyr::left_join(g2, data, by = c("link" = variable))
  } else {
    cod <- as.character(md[md$capa == unit, "cod"])
    g2 <- g %>% dplyr::select(all_of(cod)) %>%
      dplyr::rename("link" = cod) %>% 
      dplyr::mutate(link = as.numeric(link))
    data <- dplyr::left_join(g2, data %>% 
                               mutate(link = as.numeric(!!!rlang::syms(variable))), by = "link")
  }
  return(data)
}
